### PR TITLE
[SYCL] Improve handling of large reqd_work_group_size values

### PR DESCRIPTION
**Pull request by jzc**
When a kernel with a large (unsupported) work-group size is submitted, the backend would fail upon trying build the program, and throw the wrong exception (compile_program_error with `errc::build`, but it should just be a `sycl::exception` with `errc::kernel_not_supported` according to [5.8.1. Kernel attributes](https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#sec:kernel.attributes)).

https://github.com/intel/llvm/blob/b65969014f001f9730349a5caad5c2b85c9bc378/sycl/source/detail/program_manager/program_manager.cpp#L749

This PR adds a check before building to validate the `reqd_work_group_size`. 

Additionally, another bug in `sycl-post-link` was fixed. The bug occurred when a kernel was decorated with a multi-dimensional required work-group size and two dimensions had the same value. Due to how `sycl-post-link` worked, the attached metadata on the device image would include fewer dimensions than specified. (e.g. If the decorated with `reqd_work_group_size(16, 16)`, then the device image would appear as if decorated by `reqd_work_group_size(16)`.)

### CODE REVIEW

File: llvm/tools/sycl-post-link/SYCLDeviceRequirements.cpp [hunk 1]
```diff
@@ -38,13 +36,13 @@ void llvm::getSYCLDeviceRequirements(
       {"reqd_work_group_size", "reqd_work_group_size"}};
```
Comment (AlexeySachkov) [inline llvm/tools/sycl-post-link/SYCLDeviceRequirements.cpp:36]: This line should probably be removed, or otherwise we are calculating `reqd_work_group_size` requirements twice

File: llvm/tools/sycl-post-link/SYCLDeviceRequirements.cpp [hunk 2]
```diff
@@ -58,6 +56,26 @@ void llvm::getSYCLDeviceRequirements(
...
+        ReqdWorkGroupSize = NewReqdWorkGroupSize;
+      else if (!std::equal(ReqdWorkGroupSize->begin(), ReqdWorkGroupSize->end(),
+                           NewReqdWorkGroupSize.begin())) {
+        // two functions in the module have different required work group sizes
+      }
+    }
+  }
+
+  if (ReqdWorkGroupSize)
+    Requirements["reqd_work_group_size"] = std::vector<uint32_t>(
```
Comment (AlexeySachkov) [inline llvm/tools/sycl-post-link/SYCLDeviceRequirements.cpp:76]: Shouldn't it be `size_t` here? Or otherwise we are again truncating user-provided 64-bit value

File: sycl/source/detail/program_manager/program_manager.cpp [hunk 3]
```diff
@@ -2493,18 +2480,30 @@ bool doesDevSupportDeviceRequirements(const device &Dev,
...
-    std::vector<int> ReqdWGSizeVec;
+    uint64_t ReqdWGSizeAllDimsTotal = 1;
+    std::vector<uint64_t> ReqdWGSizeVec;
     int Dims = 0;
     while (!ReqdWGSize.empty()) {
-      int SingleDimSize = ReqdWGSize.consume<int>();
-      ReqdWGSizeAllDimsTotal *= SingleDimSize;
+      // The reqd_work_group_size data is stored as uint32_t's,
+      // but we'll widen the result to uint64_t.
+      uint64_t SingleDimSize = ReqdWGSize.consume<uint32_t>();
```
Comment (AlexeySachkov) [inline sycl/source/detail/program_manager/program_manager.cpp:2489]: If the spec says that `reqd-work-group-size` should support `size_t` values (KhronosGroup/SYCL-Docs#450), then we shouldn't store the attribute values in any 32-bit type at all.

Note that changing value type of the property would be an ABI-break, because we won't be able to read properties from binaries compiled using older compilers.

To workaround this, I would suggest to introduce a new `reqd_work_group_size_size_t` (or something like that) property. Compiler will always generate it, but runtime will check for both and use whichever is present, doing widening for older property. Once the next ABI-breaking window is open, we will be able to drop `_size_t` property and simply change values type of the original one.

File: sycl/source/detail/program_manager/program_manager.cpp [hunk 4]
```diff
@@ -2493,18 +2480,30 @@ bool doesDevSupportDeviceRequirements(const device &Dev,
...
+      if (auto res = multiply_with_overflow_check(ReqdWGSizeAllDimsTotal,
+                                                  SingleDimSize))
+        ReqdWGSizeAllDimsTotal = *res;
+      else
+        return sycl::exception(sycl::errc::kernel_not_supported,
+                               "Required work-group size is not supported"
+                               " (too large)");
```
Comment (AlexeySachkov) [inline sycl/source/detail/program_manager/program_manager.cpp:2496]: I wonder if this diagnostic should also be added to FE, because that is something we can check at compile-time.

I would suggest replacing "too large" with "total number of work-items requested can't fit into size_t"

File: llvm/tools/sycl-post-link/SYCLDeviceRequirements.cpp [hunk 5]
```diff
@@ -58,6 +56,26 @@ void llvm::getSYCLDeviceRequirements(
...
+      llvm::SmallVector<int64_t, 3> NewReqdWorkGroupSize;
+      for (const auto &operand : MDN->operands())
+        NewReqdWorkGroupSize.push_back(
+            ExtractIntegerFromMDNodeOperand(operand));
+      if (!ReqdWorkGroupSize)
+        ReqdWorkGroupSize = NewReqdWorkGroupSize;
+      else if (!std::equal(ReqdWorkGroupSize->begin(), ReqdWorkGroupSize->end(),
+                           NewReqdWorkGroupSize.begin())) {
+        // two functions in the module have different required work group sizes
+      }
```
Comment (AlexeySachkov) [inline llvm/tools/sycl-post-link/SYCLDeviceRequirements.cpp:71]: If the branch is empty, then we don't need it at all. We could add an assert there, but if we do so, I would also suggest to put the whole branch under something like `#ifndef NDEBUG`

File: llvm/tools/sycl-post-link/SYCLDeviceRequirements.cpp [hunk 6]
```diff
@@ -18,16 +18,14 @@
 
 using namespace llvm;
 
+auto ExtractIntegerFromMDNodeOperand(const MDOperand &operand) {
+  Constant *C = cast<ConstantAsMetadata>(operand.get())->getValue();
+  return C->getUniqueInteger().getSExtValue();
```
[the remark this answers is not in the document]
Comment (jzc) [reply llvm/tools/sycl-post-link/SYCLDeviceRequirements.cpp]: Merged with intel/sycl to incoporate changes from that PR

File: sycl/source/detail/program_manager/program_manager.cpp [hunk 7]
```diff
@@ -2475,18 +2476,21 @@ checkDevSupportDeviceRequirements(const device &Dev,
...
+    ByteArray ReqdWGSize = DeviceBinaryProperty(*(it.value())).asByteArray();
     // Drop 8 bytes describing the size of the byte array.
     ReqdWGSize.dropBytes(8);
-    uint64_t ReqdWGSizeAllDimsTotal = 1;
-    std::vector<uint64_t> ReqdWGSizeVec;
+    size_t ReqdWGSizeAllDimsTotal = 1;
+    std::vector<size_t> ReqdWGSizeVec;
     int Dims = 0;
     while (!ReqdWGSize.empty()) {
       // The reqd_work_group_size data is stored as uint32_t's,
```
Comment (AlexeySachkov) [inline sycl/source/detail/program_manager/program_manager.cpp:2490]: The comment should be updated that only old binaries store `reqd_work_group_size` as 32-bit.
We should also put `TODO`s to both runtime and the compiler to remove that extra `_size_t` property once ABI break is allowed again

File: llvm/tools/sycl-post-link/SYCLDeviceRequirements.cpp [hunk 8]
```diff
@@ -56,24 +67,23 @@ void llvm::getSYCLDeviceRequirements(
...
-      }
+      else
+        assert(std::equal(ReqdWorkGroupSize->begin(), ReqdWorkGroupSize->end(),
+                          NewReqdWorkGroupSize.begin()));
     }
   }
 
   if (ReqdWorkGroupSize)
-    Requirements["reqd_work_group_size"] = std::vector<uint32_t>(
+    Requirements["reqd_work_group_size_size_t"] = std::vector<size_t>(
```
Comment (AlexeySachkov) [inline llvm/tools/sycl-post-link/SYCLDeviceRequirements.cpp:86]: A comment is needed here that the property name should be `reqd_work_group_size`, but we couldn't change the type of existing property because we missed ABI break window; therefore, we need to rename it back (to a more obvious name) once ABI break is allowed.

Comment (AlexeySachkov) [inline llvm/tools/sycl-post-link/SYCLDeviceRequirements.cpp:86]: Unrelated to the PR (i.e. should be fixed in a separate PR if we decide to do so): Why aren't we able to construct a `PropertyValue` out of `SmallVector`? Having to create an `std::vector` kind of defeats the whole purpose of having a `SmallVector` before that

File: llvm/tools/sycl-post-link/SYCLDeviceRequirements.cpp [hunk 9]
```diff
@@ -56,24 +67,23 @@ void llvm::getSYCLDeviceRequirements(
         std::vector<uint32_t>(Values.begin(), Values.end());
   }
 
-  std::optional<llvm::SmallVector<int64_t, 3>> ReqdWorkGroupSize;
+  std::optional<llvm::SmallVector<size_t, 3>> ReqdWorkGroupSize;
```
Comment (AlexeySachkov) [inline llvm/tools/sycl-post-link/SYCLDeviceRequirements.cpp:70]: I would preserve `uint64_t` here. The tricky thing here is that we are trying to match `size_t` selected by our SYCL compiler for user's application. Strictly speaking, it may not be the same `size_t` which was selected by a compiler used to build the toolchain itself.

Having fixed-width type here at least makes it explicit and obvious about which size is supported by the toolchain and makes it independent of how toolchain is built.

File: sycl/source/detail/program_manager/program_manager.cpp [hunk 10]
```diff
@@ -2475,18 +2476,21 @@ checkDevSupportDeviceRequirements(const device &Dev,
...
+  if (ReqdWGSizeUint32TPropIt || ReqdWGSizeSizeTPropIt) {
+    bool usingSizeT = ReqdWGSizeSizeTPropIt.has_value();
+    auto it = usingSizeT ? ReqdWGSizeSizeTPropIt : ReqdWGSizeUint32TPropIt;
+
+    ByteArray ReqdWGSize = DeviceBinaryProperty(*(it.value())).asByteArray();
     // Drop 8 bytes describing the size of the byte array.
     ReqdWGSize.dropBytes(8);
-    uint64_t ReqdWGSizeAllDimsTotal = 1;
-    std::vector<uint64_t> ReqdWGSizeVec;
+    size_t ReqdWGSizeAllDimsTotal = 1;
```
Comment (AlexeySachkov) [inline sycl/source/detail/program_manager/program_manager.cpp:2486]: This is one more tricky use of a `size_t`. Once again, we are trying to match `size_t` selected by the toolchain for user application.

However, unlike with the compiler, I doubt that we can have runtime compiler with `size_t` size different from user app, because we likely pass arguments of that type between headers and the library, i.e. it is likely a part of our ABI anyway.

Therefore, I don't think that any action is required here, but personally, I would still prefer to see an explicit fixed-width data type.

File: llvm/tools/sycl-post-link/SYCLDeviceRequirements.cpp [hunk 11]
```diff
@@ -40,12 +40,10 @@ void llvm::getSYCLDeviceRequirements(
...
   // property with metadata's aspects
   constexpr std::pair<const char *, const char *> ReqdMDs[] = {
-      {"sycl_used_aspects", "aspects"},
-      {"sycl_fixed_targets", "fixed_target"},
-      {"reqd_work_group_size", "reqd_work_group_size"}};
+      {"sycl_used_aspects", "aspects"}, {"sycl_fixed_targets", "fixed_target"}};
 
   for (const auto &[MDName, MappedName] : ReqdMDs) {
-    std::set<uint32_t> Values;
+    std::set<int64_t> Values;
```
Comment (AlexeySachkov) [inline llvm/tools/sycl-post-link/SYCLDeviceRequirements.cpp:46]: Why is this change needed? Both `sycl_used_aspects` and `sycl_fixed_targets` are "internal" values of the implementation, they are not user-provided. And I don't think that any of them has to be 64-bit

File: sycl/test-e2e/OptionalKernelFeatures/large-reqd-work-group-size.cpp [hunk 12]
```diff
@@ -0,0 +1,65 @@
...
+  throws_kernel_not_supported("nd_range<3>", [] {
+    constexpr uint32_t N = 4294967295;
+    q.parallel_for<class K2>(nd_range<3>({N, N, N}, {N, N, N}),
+                             [=](auto)
+                                 [[sycl::reqd_work_group_size(N, N, N)]] {});
+  });
+
+  // TODO: Due to truncation issues, this test cannot pass yet.
+  // Enable this test once fixed.
+#ifndef NO_RANGE_GREATER_THAN_UINT32_MAX
```
Comment (AlexeySachkov) [inline sycl/test-e2e/OptionalKernelFeatures/large-reqd-work-group-size.cpp:56]: Comment and corresponding `#ifndef` macro should also be removed from the test

File: llvm/tools/sycl-post-link/SYCLDeviceRequirements.cpp (hunk 11 above)
Comment (jzc) [reply llvm/tools/sycl-post-link/SYCLDeviceRequirements.cpp]: I think it might have been necessary before merging with #10614; now that that PRs changes are integrated with this PR it should be unnecessary.

**Review by AlexeySachkov — APPROVED** [review]
`llvm/` part LGTM.

@intel/sycl-language-enabling-triage: FYI

File: sycl/test-e2e/OptionalKernelFeatures/large-reqd-work-group-size.cpp [hunk 13]
```diff
@@ -0,0 +1,62 @@
...
+  }
+  std::cout << "fail: " << test_name << "\n"
+            << "No exception thrown\n";
+  ++n_fail;
+  return;
+}
+
+int main(int argc, char *argv[]) {
+  throws_kernel_not_supported("nd_range<1>", [] {
+    constexpr uint32_t N = 4294967295;
```
Comment (maarquitos14) [inline sycl/test-e2e/OptionalKernelFeatures/large-reqd-work-group-size.cpp:37]: Could we use `std::numeric_limits<T>::max()` rather than the value?

File: sycl/test-e2e/OptionalKernelFeatures/large-reqd-work-group-size.cpp [hunk 14]
```diff
@@ -0,0 +1,62 @@
...
+
+  throws_kernel_not_supported("nd_range<3>", [] {
+    constexpr uint32_t N = 4294967295;
+    q.parallel_for<class K2>(nd_range<3>({N, N, N}, {N, N, N}),
+                             [=](auto)
+                                 [[sycl::reqd_work_group_size(N, N, N)]] {});
+  });
+
+  throws_kernel_not_supported("uint32_max+2", [] {
+    constexpr uint64_t N = 4294967297;
```
Comment (maarquitos14) [inline sycl/test-e2e/OptionalKernelFeatures/large-reqd-work-group-size.cpp:56]: Should we try if this value throws for `uint32_t`?

File: llvm/tools/sycl-post-link/SYCLDeviceRequirements.cpp [hunk 15]
```diff
@@ -69,6 +67,27 @@ void llvm::getSYCLDeviceRequirements(
...
   }
 
+  std::optional<llvm::SmallVector<uint64_t, 3>> ReqdWorkGroupSize;
+  for (const Function &F : MD.getModule()) {
+    if (const MDNode *MDN = F.getMetadata("reqd_work_group_size")) {
+      llvm::SmallVector<size_t, 3> NewReqdWorkGroupSize;
+      for (size_t I = 0, E = MDN->getNumOperands(); I < E; ++I)
+        NewReqdWorkGroupSize.push_back(
+            ExtractUnsignedIntegerFromMDNodeOperand(MDN, I));
+      if (!ReqdWorkGroupSize)
```
Comment (KornevNikita) [inline llvm/tools/sycl-post-link/SYCLDeviceRequirements.cpp:77]: It looks like now we save only the first occurrence of the `reqd_work_group_size` metadata, isn't it?

**Review by KornevNikita — APPROVED** [review]

File: sycl/test-e2e/OptionalKernelFeatures/large-reqd-work-group-size.cpp (hunk 14 above)
Comment (maarquitos14) [reply sycl/test-e2e/OptionalKernelFeatures/large-reqd-work-group-size.cpp:56]: Yeah, that was what I meant. So what is the expected behavior if we use `uint32_t` as type with the same value? Is it undefined behavior?

**Review by maarquitos14 — APPROVED** [review]
LGTM.

**Comment by dm-vodopyanov** [comment]
@jzc can you please fix the failure in post-commit?

https://github.com/intel/llvm/actions/runs/5819408460/job/15777743360
